KTB-17-send-questions-function

DenisGubernatorov opened the pull request:
1) Сообщение о том, что все слова выучены вынесено в отдельную константу, для использования и в консольном режиме в боте 2) updateId сделан свойством класса, для использования в методе отправки вопросов 3) Кнопка статистики, из аргументов метода убрана передача trainer, так как он является свойством класса, с соответствующей видимость 4) Текст для запроса колбэка сделан константой
5) Реализован метод поочередной отправки вопросов, с ожиданием ответа на каждый вопрос 6) в TelegramBotService метод создания getHttpRequest вынесен в отдельный метод 7) Реализован класс формирующий тело запроса с вопросом и вариантами ответов в виде кнопок. Кнопки располагаются по вертикали.

DenisGubernatorov committed:
KTB-17-send-questions-function
1) Сообщение о том, что все слова выучены вынесено в отдельную константу, для использования и в консольном режиме в боте
2) updateId сделан свойством класса, для использования в методе отправки вопросов
3) Кнопка статистики, из аргументов метода убрана передача trainer, так как он является свойством класса, с соответствующей видимость
4) Текст для запроса колбэка сделан константой
5) Реализован метод поочередной отправки вопросов, с ожиданием ответа на каждый вопрос
6) в TelegramBotService метод создания getHttpRequest вынесен в отдельный метод
7) Реализован класс формирующий тело запроса с вопросом и вариантами ответов в виде кнопок. Кнопки располагаются по вертикали.

DenisGubernatorov commented:
1. Смущает то, что анимация нажатия кнопок в телеге длится около 5 секунд.  Я думал что это связано с тем, что у меня есть допполнительный цикл для ожидания ответа на вопрос. Но и для кнопки статистики, где ничего не ждет, такая же ситуация
![image](https://github.com/user-attachments/assets/61e00d36-3758-4c0c-b84d-17af0e5b6560)

DenisGubernatorov commented:
Не очень понятно с ТЗ. Вроде сначала в описании упоминается метод sendQuestion, а потом, в деталях реализации речь про senMessage. Просто текущий sendMessage реализован через Get запрос и для запросов с кнопками он не подходит. Реализовал именно метод sendQuestion
![image](https://github.com/user-attachments/assets/e76807f3-987c-4cc1-9afa-6c30203e3560)

dzenonce commented:
> 1. Смущает то, что анимация нажатия кнопок в телеге длится около 5 секунд.  Я думал что это связано с тем, что у меня есть допполнительный цикл для ожидания ответа на вопрос. Но и для кнопки статистики, где ничего не ждет, такая же ситуация
>        ![image](https://private-user-images.githubusercontent.com/72883929/378987265-61e00d36-3758-4c0c-b84d-17af0e5b6560.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2ODAzNTUsIm5iZiI6MTcyOTY4MDA1NSwicGF0aCI6Ii83Mjg4MzkyOS8zNzg5ODcyNjUtNjFlMDBkMzYtMzc1OC00YzBjLWI4NGQtMTdhZjBlNWI2NTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDEwNDA1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlNmJjYzU4ZGZhY2U5YzNlYWMxNzAyNzNmMDQyNGRlZDk2NWIyYTk0ZmM0ODIzOTAzNWE1YWJmOWRlZWNlMjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DDGewbE3OpnwQvHERYSUt-0QGgkZIJp77tbYZ9gbXOw)

Замечательно что обратил на это внимание, эта ситуация связана с самим ТГ, в api есть метод обработки этой ситуации)

danelyan commented on src/main/kotlin/Telegram.kt:
```diff
@@ -51,21 +52,40 @@ private fun workByCommand(
+                    updateId++
+                    noAnswer = false
+                }
+            } while (noAnswer)
```
Так лучше не делать (я про цикл), подробнее описал в видео-ревью:
https://drive.google.com/file/d/1pnnyAi3xAOa64YfxPcUYH4h4vbvJs0le/view?usp=sharing

Если будут какие-то вопросы, напиши сюда в тред или [мне в личку ](https://t.me/fridonium)

danelyan replied:
> Смущает то, что анимация нажатия кнопок в телеге длится около 5 секунд. Я думал что это связано с тем, что у меня есть допполнительный цикл для ожидания ответа на вопрос. Но и для кнопки статистики, где ничего не ждет, такая же ситуация

Думаю это может быть связано с тем, что АПИ телеграма отправляет тебя в RateLimit. То есть, ограничивает скорость ответа на твои запросы. Попробуй позапускать в дебаг-режиме и посмотри, где ты находишься на момент моргания кнопки

DenisGubernatorov replied:
> Так лучше не делать (я про цикл), подробнее описал в видео-ревью: https://drive.google.com/file/d/1pnnyAi3xAOa64YfxPcUYH4h4vbvJs0le/view?usp=sharing
> 
> Если будут какие-то вопросы, напиши сюда в тред или [мне в личку ](https://t.me/fridonium)

Ну просто когда реализовывали консольный режим я из задачи понял, что надо все слова показывать. Ну и решил придерживаться для бота.
![image](https://github.com/user-attachments/assets/67747155-beb8-4a2f-beae-b965b273d952)

danelyan commented:
> Не очень понятно с ТЗ. Вроде сначала в описании упоминается метод sendQuestion, а потом, в деталях реализации речь про senMessage. Просто текущий sendMessage реализован через Get запрос и для запросов с кнопками он не подходит. Реализовал именно метод sendQuestion

Тут терминологическая путаница, под понятием `для метода sendMessage` подразумевается метод API, а не наш метод. Наших методов может быть несколько, и они используют метод `sendMessage`. В каких то случаях через GET, в каких-то POST.

DenisGubernatorov committed:
KTB-17-send-questions-function
1) Изменение логики. При выборе режима изучения слов бот теперь показывает только одно случайно выбранное слово для изучения, а не все неизученные. Соответственно доработан и консольный вариант.
2) updateId сделал опять локальной переменной
3) trainer сделан локальной переменной. Протянута передача в методы, где он используется
4) В TelegramBotService добавлен метод оповещающий бота о том, что нажатие на кнопку получено. Нужно для того, что бы избежать долгой анимации нажатия кнопки в боте
